### PR TITLE
implementation likes look up by tweet

### PR DESCRIPTION
**Pull request by kazdevl**
## TODO
- [x] define Options
- [x] define Response
- [x] create client method
- [x] test above

## REF
- [api reference](https://developer.twitter.com/en/docs/twitter-api/tweets/likes/api-reference/get-tweets-id-liking_users)

### CODE REVIEW

File: likes_lookup.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+	"net/http"
+)
+
+func likesLookUpUsers(ctx context.Context, c *client, id string, opt ...*LikesLookUpByTweetOpts) (*LikesLookUpByTweetResponse, error) {
+	// check id
+	if len(id) == 0 {
+		return nil, errors.New("likes look up by tweet: id parameter is required")
+	}
+	LikesLookUpByTweet := timeline + "?id=" + id + "/tweets"
```
Comment (sivchari) [inline likes_lookup.go:16]: This api endpoint is `https://api.twitter.com/2/tweets/:id/liking_users`.
But this line's endpoint is `https://api.twitter.com/2/users?id={id}/tweets`
(timeline = "https://api.twitter.com/2/users")

you may mistake implementation.

**Review by sivchari — CHANGES_REQUESTED** [review]

**Comment by sivchari** [comment]
Please add test.

File: client.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ type Client interface {
 	SearchSpaces(ctx context.Context, query string, opt ...*SearchSpacesOption) (*SearchSpacesResponse, error)
 	UserMentionTimeline(ctx context.Context, id string, opt ...*UserMentionTimelineOpts) (*UserMentionTimelineResponse, error)
 	UserTweetTimeline(ctx context.Context, id string, opt ...*UserTweetTimelineOpts) (*UserTweetTimelineResponse, error)
+	LikesLookUpUsers(ctx context.Context, id string, opt ...*LikesLookUpUserOpts) (*LikesLookUpUserResponse, error)
```
Comment (sivchari) [inline client.go:40]: Please order of alphabet.

Comment (sivchari) [inline client.go:40]:
```suggestion
	LookUpLikesByUsers(ctx context.Context, tweetID string, opt ...*LikesLookUpUserOpts) (*LikesLookUpUserResponse, error)
```

File: likes_lookup.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	"fmt"
+	"net/http"
+)
+
+func likesLookUpUsers(ctx context.Context, c *client, id string, opt ...*LikesLookUpUserOpts) (*LikesLookUpUserResponse, error) {
+	// check id
+	if len(id) == 0 {
+		return nil, errors.New("likes look up by tweet: id parameter is required")
+	}
+	LikesLookUpByTweet := tweetLookUp + "/" + id + "/liking_users"
```
Comment (sivchari) [inline likes_lookup.go:16]: Please don't use `tweetLookUp`.
This endpoint is `likes API`. Not `tweet look up`.

File: likes_lookup.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+func likesLookUpUsers(ctx context.Context, c *client, id string, opt ...*LikesLookUpUserOpts) (*LikesLookUpUserResponse, error) {
+	// check id
+	if len(id) == 0 {
+		return nil, errors.New("likes look up by tweet: id parameter is required")
```
Comment (sivchari) [inline likes_lookup.go:14]: This error message is strange.
```suggestion
		return nil, errors.New("look up likes by users: tweet id parameter is required")
```

File: likes_lookup.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	LikesLookUpByTweet := tweetLookUp + "/" + id + "/liking_users"
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, LikesLookUpByTweet, nil)
+	if err != nil {
+		return nil, fmt.Errorf("likes look up by tweet new request with ctx: %w", err)
```
Comment (sivchari) [inline likes_lookup.go:20]:
```suggestion
		return nil, fmt.Errorf("look up likes by users: new request with ctx: %w", err)
```

File: likes_lookup.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.bearerToken))
+
+	var topt LikesLookUpUserOpts
+	switch len(opt) {
+	case 0:
+		// do nothing
+	case 1:
+		topt = *opt[0]
+	default:
+		return nil, errors.New("likes look up by tweet: only one option is allowed")
```
Comment (sivchari) [inline likes_lookup.go:31]:
```suggestion
		return nil, errors.New("look up likes by users:  only one option is allowed")
```

File: likes_lookup.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	topt.addQuery(req)
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("likes look up by tweet response: %w", err)
```
Comment (sivchari) [inline likes_lookup.go:37]:
```suggestion
		return nil, fmt.Errorf("look up likes by users response: %w", err)
```

File: likes_lookup.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	defer resp.Body.Close()
+
+	var usersWhoLiked LikesLookUpUserResponse
+	if err := json.NewDecoder(resp.Body).Decode(&usersWhoLiked); err != nil {
+		return nil, fmt.Errorf("likes look up by tweet decode: %w", err)
```
Comment (sivchari) [inline likes_lookup.go:43]:
```suggestion
		return nil, fmt.Errorf("look up likes by users decode: %w", err)
```

File: likes_lookup.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	}
+	if resp.StatusCode != http.StatusOK {
+		return &usersWhoLiked, &HTTPError{
+			APIName: "likes look up by tweet",
```
Comment (sivchari) [inline likes_lookup.go:47]:
```suggestion
			APIName: "look up likes by users",
```

File: likes_lookup_test.go [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+		opt    []*gotwtr.LikesLookUpUserOpts
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *gotwtr.LikesLookUpUserResponse
+		wantErr bool
+	}{
+		{
+			name: "200 success with No Optional fileds",
```
Comment (sivchari) [inline likes_lookup_test.go:27]:
```suggestion
			name: "200 success with no optional fields",
```

File: likes_lookup_test.go [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+						ID:       "1526228120",
+						Name:     "Twitter Data",
+						UserName: "TwitterData",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "200 success with Optional fileds",
```
Comment (sivchari) [inline likes_lookup_test.go:78]:
```suggestion
			name: "200 success with optional fields",
```

File: likes_lookup_test.go [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			c := gotwtr.New("test-key", gotwtr.WithHTTPClient(tt.args.client))
+			got, err := c.LikesLookUpUsers(tt.args.ctx, "tweet_id", tt.args.opt...)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("LikesLookUpUsers() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if diff := cmp.Diff(got, tt.want); diff != "" {
+				t.Errorf("LikesLookUpUsersm() mismatch (-want +got):\n%s", diff)
```
Comment (sivchari) [inline likes_lookup_test.go:226]: typo.

**Review by sivchari — CHANGES_REQUESTED** [review]
Please check review.

File: likes_lookup_test.go [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/sivchari/gotwtr"
+)
+
+func Test_lookUpUsersWhoLiked(t *testing.T) {
+	type args struct {
```
Comment (sivchari) [inline likes_lookup_test.go:15]: t.Parallel()

**Review by sivchari — CHANGES_REQUESTED** [review]

File: likes_lookup_test.go (hunk 13 above)
Comment (kazdevl) [reply likes_lookup_test.go:15]: Is your intention to add `t.Parallel()` to the subtest? If so, I have already added it.
If your intention is to add `t.Parallel()` to top-level test, I think it is unnecessary because in the other top-level tests, `t.Parallel()` is not present.

Comment (sivchari) [reply likes_lookup_test.go:15]: > I think it is unnecessary because in the other top-level tests, t.Parallel() is not present.

If this is reason, it's not good reason.
I'll add `t.Parallel()` to all top-level-test, soon.
If you have another reason, please describe.

Comment (kazdevl) [reply likes_lookup_test.go:15]: I thought it would be more appropriate to add `t.Parallel()` to all top-level-tests together after merging this PullRequest, intead of adding `t.Parallel()` to a top-level-test in this PullRequest, in terms of the semantics of that commit that you try to do.

Comment (kazdevl) [reply likes_lookup_test.go:15]: above `commit that you try to do` is intended for `I'll add t.Parallel() to all top-level-test, soon.`

Comment (sivchari) [reply likes_lookup_test.go:15]: > I thought it would be more appropriate to add t.Parallel() to all top-level-tests together after merging this PullRequest, intead of adding t.Parallel() to a top-level-test in this PullRequest, in terms of the semantics of that commit that you try to do.

I don't agree with above comment.
A future fix is ​​to add to tests that don't currently support `t.Parallel ()`.
The problem this time is that there is no `t.Parallel ()` at the top level of this PR.
In other words, it is completely different from intentionally not supporting `t.Parallel ()` this time.

**Review by sivchari — APPROVED** [review]
LGTM

**Comment by sivchari** [comment]
please change.
